packages/react-dom/src/__tests__/ReactMount-test.js

```diff
@@ -173,12 +482,632 @@ describe('ReactMount', () => {
+    );
+  });
+
+  it('should warn when hydrate inserts an element after a comment node (insertion diff)', () => {
+    const div = document.createElement('div');
+    div.innerHTML = '<div><!-- a comment --></div>';
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <span />
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <span> in <div>.\n\n' +
+        '  <div>\n' +
+        '    <!-- a comment -->\n' +
+        '+   <span />\n' +
+        '  </div>\n\n' +
+        '    in span (at **)\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn when hydrate replaces an element after a comment node (replacement diff)', () => {
+    const div = document.createElement('div');
+    div.innerHTML = '<div><!-- a comment <-- --&gt; > --><div></div></div>';
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <span />
+          <div />
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <span> in <div>.\n\n' +
+        '  <div>\n' +
+        '    <!-- a comment <-- --&gt; > -->\n' +
+        '-   <div />\n' +
+        '+   <span />\n' +
+        '  </div>\n\n' +
+        '    in span (at **)\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn when hydrate inserts an element after a non-typical node (insertion diff)', () => {
+    // A non-typical node is a node that is not typically seen in the DOM: not an element, text, or comment.
+    // This is an artificial test case to check how we print non-typical nodes if they somehow end up in the DOM.
+    const xml = document.createElement('xml');
+    xml.appendChild(
+      document.createProcessingInstruction(
+        'dom-processing-instruction',
+        'content > ',
+      ),
+    );
+
+    expect(() => ReactDOM.hydrate(<div />, xml)).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <div> in <xml>.\n\n' +
+        '  <xml>\n' +
+        '    <?dom-processing-instruction content &gt; ?>\n' +
+        '+   <div />\n' +
+        '  </xml>\n\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn with special characters in the JSX string output', () => {
+    const div = document.createElement('div');
+    const markup = ReactDOMServer.renderToString(
+      <div>{'SSRMismatchTest special characters: \'"\t\n'}</div>,
+    );
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <span />
+          {'SSRMismatchTest special characters: \'"\t\n'}
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      // TODO: Currently, the special characters in the JSX string output are not escaped, the output looks invalid.
+      'Warning: Expected server HTML to contain a matching <span> in <div>.\n\n' +
+        '  <div data-reactroot="">\n' +
+        "-   {'SSRMismatchTest special characters: '\"\t\n'}\n" +
```

Comment:
TODO: Currently, the special characters in the JSX string output are not escaped, the output looks invalid.

Despite there's code that should escape them (`escapeNonPrintableCharacters`). Need to look into it once again.

Reply:
> TODO: Currently, the special characters in the JSX string output are not escaped, the output looks invalid.
>
> Despite there's code that should escape them (`escapeNonPrintableCharacters`). Need to look into it once again.

Fixed in https://github.com/facebook/react/pull/13602/commits/c0b778097f0af2c950e8de5bac3abb244e195a08